---
title: "TCGA example"
output: html_notebook
---
First steps are to load the dependencies and source the function files

```{r}
library(tidyverse)
library(org.Hs.eg.db)
library(AnnotationDbi)

setwd("/home/jacobo/Documents/05_DeconvolutionReview/02_TCGA/")
source(file = "/home/jacobo/Documents/05_DeconvolutionReview/R/functions.R")
```

Then we load the data and metadata, with some filtering in the metadata too.

```{r message=FALSE, warning=FALSE, include=FALSE}
tcga_raw <- read_tsv("01_Input/TCGA_TPM_PDAC.tsv") %>%
  rename_all(~str_replace_all(.,"-", ".")) # JADE change this automatically after, so better start with them like this

metadata <- read_tsv("01_Input/TCGA-CDR-SupplementalTableS1.tsv", col_types = cols(`...1` = col_skip())) %>%
  mutate(bcr_patient_barcode = str_replace_all(bcr_patient_barcode, "-", ".")) %>%
  mutate(OS.time = as.double(OS.time),
         PFI.time = as.double(PFI.time)) %>%
  dplyr::filter(bcr_patient_barcode %in% colnames(tcga_raw)) %>% 
  mutate_all( ~ replace(.,  . %in% c("#N/A","[Not Applicable]", "[Not Available]"), NA)) %>%
  select_if(~sum(!is.na(.)) > 0)

metadata_pdac_death <- metadata %>% dplyr::filter(DSS == 1)

tcga <- tcga_raw %>% 
    dplyr::select(c(gene, metadata_pdac_death$bcr_patient_barcode))
```

Next we need to prepare the TCGA data so we are able to run ICA in it. For that 

#1 Filter the sexual chromosomes genes to avoid getting a sex component
```{r message=FALSE, warning=FALSE, include=FALSE}
annot <- AnnotationDbi::select(org.Hs.eg.db, keys=tcga$gene, columns=c("ENSEMBL", "MAP"), keytype="SYMBOL")
to_keep <- dplyr::filter(annot, !grepl('X|Y', MAP)) %>%
  drop_na(MAP) %>%
  dplyr::select("SYMBOL") %>%
  distinct()

tcga_gn <- dplyr::filter(tcga, gene %in% to_keep$SYMBOL)
```

#2 Mean center genewise and selection of half most variable genes. Finally save an .RDS object as a checkpoint

```{r}
tcga__ <- tcga_gn %>% pivot_longer(-gene, 'variable', 'value') %>%
  pivot_wider(id_cols = variable, names_from = gene) %>%
  mutate(across(where(is.numeric), ~ . - mean(.))) %>%
  pivot_longer(-variable, "gene",  "value") %>% 
  pivot_wider(id_cols = gene, names_from = variable)

abdv <- apply(tcga__ %>% dplyr::select(-gene), 1, function(x) { #! could be piped with  add_row(variable = "mean", !!! colMeans(.[-1])) when transposed
  sum(
    abs(
      x - mean(x)
    )
  ) / length(x)
})

median_abdv <- median(abdv)
tcga_p <- tcga__[abdv > median_abdv, ]

# plots with A1CF as example gene
pivot_longer(tcga_gn, -gene) %>% 
  dplyr::filter(gene == "A1CF") %>% 
  ggplot(aes(x = value)) + geom_density()

pivot_longer(tcga_p, -gene) %>% 
  dplyr::filter(gene == "A1CF") %>% 
  ggplot(aes(x = value)) + geom_density()

write_rds(tcga_p, file = "01_Input/tcga_icaready.RDS")
```

Now that the data is processed  there are two different ways ICA_toolkit can be used:

#1 Find the most robust number of componenst (i.e. the number in which less components are caused by outliers).

This is recommended for a unsupervised analysis. This allow the researcher to choose the number of components more robust for your data. This way you minimize the components due to single or little samples (outliers).

## CAREFUL: Multithreading to be implemented. this example takes around 7 mins to run
```{r}
tcga_p <- read_rds("01_Input/tcga_icaready.RDS")

boot_res <- Boot_ICA(expression = tcga_p, df_id  =  "gene", range.comp = 2:20, iterations = 20, seed = 0)

ggplot(boot_res) + aes(x = nc, y = correlation, fill = bootstrap) +
    geom_boxplot(width = 0.5) +
    labs(y = "absolute pearson correlation", x = "number of components") +
    coord_cartesian(ylim = c(0.8, 1)) +
    theme_bw() +
    rotate_x_text(45)
```
The idea is to then take the most robust number fo components (with the best correlations with the bootstrapped dataframes), but that as well allow us to keep the most components the better. Here I would choose 13 components.


#2 Find Molecular signature
When we know the existence of a factor with consequences at the gene expression 
level we can use ICA_toolkit to produce a molecular signature of this factor and
then be able to apply it to other datasets. For that first we directly generate
ICAs over a range of components and then we will find the component that best
represent the factor.

the generation of ICAs over a range of components can be done with Range_ICA()

```{r}
tcga_p <- read_rds("01_Input/tcga_icaready.RDS")
icas_list <- Range_ICA(tcga_p, "gene", 2:20)
```

Once we have ICAs over the range of components we can for instance look for the
best component representing "OS.time" and "PFI.time". In this case we use just 
the patients with a confirmed PDAC caused dead.

This is done with the Best_nc() function.

```{r}
Best_nc(icas_list,
        range.comp = 2:20,
        metadata = metadata_pdac_death,
        metadata_id = "bcr_patient_barcode",
        vars = "OS.time",
        is.categorical = F)
```

Then we can check the graph and choose the best ICA space based in the number of components. In our case we are interested in a space where a single component represent the best these two variables. For that reason we chose the ICA space for nc = 18 components, where we are particularly interested in the IC. 17.


# Exploration of the component space:

So idependently on the method of choice, after choosing the ICA space we need to re-run Range_ICA, but this time only with the elected number of components to check the distribution of the rest of variables across this space, and try to figure out what each component is representing.

```{r}
best_ica <- Range_ICA(tcga_p, "gene", 18)
```

Now its time to analyze the ICA space we have chosen. For that the first step is to plot the components distribution to discriminate those components that could be caused by outliers. If annotations are provided as a dataset (with the same sample order) one plot will be generated per column contrastimg their distribution compared to the component space. Make sure that your continuous labels are doubles!

```{r}
annotations <- metadata_pdac_death %>%
  dplyr::select("bcr_patient_barcode", "OS.time", "gender")

Sampleweights_distribution(ica = best_ica,
                    df = annotations,
                    df_id = "bcr_patient_barcode")
```
Now its the time to check the whole component space. For that we implemented a default function named ICA_explorator() to generate correlation plots with the data of your choice. We recommend to do a first test with the technical information available, as ICA can easily pick this kind of variability into a component.

In this case we dont have such information, so we directly check the rest of the clinical data.

```{r message=FALSE, warning=FALSE, include=FALSE}
#! Fix conditional testing of vars with not enough datapoints
#! Fix ploting size for ncomponents close to 20
## clinical data
metadata_cont <- c("age_at_initial_pathologic_diagnosis",
                   "initial_pathologic_dx_year",
                   "new_tumor_event_dx_days_to", "OS.time", "PFI.time", "DFI.time")
# metadata_disc <- c("gender", "race", "histological_type", "ajcc_pathologic_tumor_stage",
#                    "histological_grade", "vital_status", "tumor_status",
#                    "new_tumor_event_type", "new_tumor_event_site",
#                    "treatment_outcome_first_course", "OS", "DSS", "DFI")

metadata_disc <- c("gender", "race", "histological_type", "ajcc_pathologic_tumor_stage",
                   "histological_grade")

### find IC responsible
clinical_data_plots <- ICA_explorator(
  ica = best_ica,
  df = metadata_pdac_death,
  df_cont = metadata_cont,
  df_disc = metadata_disc,
  df_id = "bcr_patient_barcode")

```

This Function returns a list containing 2 ggplot objects you can manipulate to save as you desire
```{r,  fig.width=10, fig.height=3}
clinical_data_plots$cont  
```

```{r, fig.width=10, fig.height=8}
clinical_data_plots$disc
```

This function can take whatever kind of continuous/discrete data and compare it to your weights based in an id column. For instance you could use this to try to relate the space we have (and specially IC. 17) to the estimated activity of the 50% most variable TFs

```{r,  fig.width=8, fig.height=15}
library(dorothea)
# Generate TF act 
### Select the regulon data
data(dorothea_hs_pancancer, package = "dorothea")
regulons <- dorothea_hs_pancancer %>%
  dplyr::filter(confidence %in% c("A", "B","C"))

## VIPER
minsize = 5
ges.filter = FALSE

tf_act__<- dorothea::run_viper(column_to_rownames(tcga, "gene"), regulons,
                                          options =  list(minsize = minsize, eset.filter = ges.filter, 
                                                          cores = 1, verbose = FALSE, nes = TRUE))

# Format for analysis
tf_act_ <- as_tibble(tf_act__, rownames = "TF") %>%
  pivot_longer(cols = -TF, names_to = "bcr_patient_barcode") %>%
  pivot_wider(id_cols = c(bcr_patient_barcode), names_from = TF)

# get 50 most variable
tf_50_var <- summarise(tf_act_, across(where(is.double), var)) %>%
  pivot_longer(cols = everything(), names_to = "TF", values_to = "var") %>%
  dplyr::arrange(desc(var)) %>% dplyr::slice_head(n=50)

tf_act <- tf_act_ %>% dplyr::select(bcr_patient_barcode, all_of(tf_50_var$TF))
### find IC responsible
TF_plots <- ICA_explorator(
  ica = best_ica,
  df = tf_act,
  df_cont = tf_50_var$TF,
  df_id = "bcr_patient_barcode")

TF_plots$cont
```

As well, you could build special functions for analyzing other kind of data. For instance we implemented a special function to analyze the different deconvolution methods and signatures segregated by CT of the output of  the GemDeCan deconvolution package (https://github.com/VeraPancaldiLab/GEMDeCan/). TO BE FIXED


```{r, fig.width=10, fig.height=8}
# need to fix!
tcga_decon <- read_rds("01_Input/TCGA_PAAD_curated_deconvolutions_m.RDS") %>%
  as_tibble(rownames = "samples") %>% 
  mutate(samples = str_replace_all(samples, "-", ".")) %>% 
  column_to_rownames("samples")

res <- Corr_by_ct(best_ica[["A"]], tcga_decon)
```

We also offer an additional function to quickly explore any individual correlation you would like more in depth, specially related to the particular component IC.17. This is Sampleweights_indepth(), and can take continuous or discrete variables. it returns a ggplot object so you can choose to plot it as you like. You can use any ggplot function or strategy in these including coloring for additional factors
```{r}
ic17_gender <- Sampleweights_indepth(ica = best_ica,
                      interest_IC = "IC.17",
                      df = metadata_pdac_death,
                      df_id = "bcr_patient_barcode",
                      var = "gender",
                      test = "ttest")

ic17_gender + 
  geom_violin(aes(fill = gender), position=position_dodge(0.8), width=0.5) +
  geom_boxplot(position=position_dodge(0.8), width=0.1) +
  theme_classic()

ic17_OS <- Sampleweights_indepth(ica = best_ica,
                      interest_IC = "IC.17",
                      df = metadata_pdac_death,
                      df_id = "bcr_patient_barcode",
                      var = "OS.time",
                      test = "spearman")

ic17_OS + geom_point() + geom_smooth(method=lm) +
  scale_colour_gradientn(colours = terrain.colors(10)) +
  theme_bw()

```
You can use any ggplot function or strategy in these including coloring for additional factors
```{r}
ic17_OS <- Sampleweights_indepth(ica = best_ica,
                      interest_IC = "IC.17",
                      df = metadata_pdac_death,
                      df_id = "bcr_patient_barcode",
                      var = "OS.time")

ic17_OS + geom_point(aes(color = gender)) + geom_smooth(aes(color = gender), method=lm, se = F) +
  theme_bw()

```


Its very important to check the gene weights of the component space, as it comes in thousands, not in hundreds like the samples. Directly testing these with ssGSEA, or just checking the 50 genes at each side of the tail can give a nice overview of what is behind of the component. For that we provide the function PlotGeneWeights. The functions return two plots on a list, so you can modify or orgize it as you want.

```{r, fig.width=10, fig.height=8}
gw <- PlotGeneWeights(ica = best_ica,
                      interest_IC = "IC.17",
                      expression = tcga_gn,
                      df_id = "gene",
                      n_genes  = 25,
                      column_annotation = NA)

ggarrange(gw[["densplot"]] + rremove("xylab"), gw[["heatmap"]], heights = c(1.5, 10), widths = c(0.2,1),
          labels = c("A", "B"),
          ncol = 2, nrow = 1)
```

Its also recomended to perform in these supervised dimensionality reduction techniques as GSVA. Here is an example checking IC.17

```{r, fig.width=10, fig.height=8}
library(msigdbr)
library(GSVA)

all_genesets <- msigdbr("Homo sapiens")
use_genesets <- dplyr::filter(all_genesets, gs_subcat %in% c("CP:REACTOME"))
msigdbr_list <- split(x = use_genesets$gene_symbol, f = use_genesets$gs_name)
msigdb_descriptions <- use_genesets[c("gs_name", "gs_description")] %>%
  unique() %>% column_to_rownames("gs_name")

gsvaRes <- gsva(data.matrix(best_ica$S), msigdbr_list, min.sz = 15)

# best for IC.17
gsvaTop <- as_tibble(gsvaRes, rownames = "gs_name") %>%  mutate(the_rank = rank(IC.17, ties.method = "random")) %>% 
  dplyr::filter(the_rank < 26 | the_rank > (nrow(gsvaRes)-26)) %>% 
  dplyr::mutate(gs_name = fct_reorder(gs_name, IC.17)) %>% mutate(sign = case_when(IC.17 < 0 ~ "neg", IC.17 > 0 ~ "pos")) %>%
  dplyr::select(gs_name, IC.17, sign)

ggplot(gsvaTop, aes(x = IC.17, y = gs_name, color = sign)) + 
  geom_point() +
  ggtitle("Best Reactome gene sets IC.17") +
  xlab("NES") +
  theme_bw() +
  rremove("legend") +
  rremove("ylab")
```
